src/AutoRest.CSharp/LowLevel/AutoRest/DpgOutputLibrary.cs

```diff
@@ -12,29 +13,30 @@ namespace AutoRest.CSharp.Output.Models.Types
     internal class DpgOutputLibrary : OutputLibrary
     {
         private readonly IReadOnlyDictionary<InputEnumType, EnumType> _enums;
-        private readonly IReadOnlyDictionary<InputModelType, ModelTypeProvider> _models;
-        private readonly TypeFactory _typeFactory;
+        private readonly IReadOnlyDictionary<InputModelType, Func<ModelTypeProvider>> _modelFactories;
 
+        public TypeFactory TypeFactory { get; }
         public IEnumerable<EnumType> Enums => _enums.Values;
-        public IEnumerable<ModelTypeProvider> Models => _models.Values;
+        public IEnumerable<ModelTypeProvider> Models => _modelFactories.Values.Select(f => f());
         public IReadOnlyList<LowLevelClient> RestClients { get; }
         public ClientOptionsTypeProvider ClientOptions { get; }
 
-        public DpgOutputLibrary(Func<TypeFactory, IReadOnlyDictionary<InputEnumType, EnumType>> enumsFactory, Func<TypeFactory, IReadOnlyDictionary<InputModelType, ModelTypeProvider>> modelsFactory, Func<TypeFactory, IReadOnlyList<LowLevelClient>> restClientsFactory, ClientOptionsTypeProvider clientOptions)
+        public DpgOutputLibrary(IReadOnlyDictionary<InputEnumType, EnumType> enums, IReadOnlyDictionary<InputModelType, Func<ModelTypeProvider>> modelFactories, IReadOnlyList<LowLevelClient> restClients, ClientOptionsTypeProvider clientOptions)
```

Comment:
So here we break the circular dependency between `ModelTypeProvider`, `TypeFactory` and `DpgOutputLibrary`?

Reply:
Kind of. I hope that when we move MPG on the CADL rails, we can fix this completely.